projects/RabbitMQ.Client/client/impl/WireFormatting.cs

```diff
@@ -700,39 +718,75 @@ public static int WriteShort(Span<byte> span, ushort val)
         }
 
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        public static unsafe int WriteShortstr(Span<byte> span, string val)
+        public static int WriteShortstr(Span<byte> span, string val)
         {
+            if (string.IsNullOrEmpty(val))
+            {
+                span[0] = 0;
+                return 1;
+            }
+
             int maxLength = span.Length - 1;
             if (maxLength > byte.MaxValue)
             {
                 maxLength = byte.MaxValue;
             }
-            fixed (char* chars = val)
-            fixed (byte* bytes = &span.Slice(1).GetPinnableReference())
+
+            return WriteShortStrImpl(span, val, maxLength);
+        }
+
+        private static int WriteShortStrImpl(Span<byte> span, string val, int maxLength)
```

Comment:
Now I'm hooked =) Was it beneficial to extract this part in it's own method? (I assume this won't be inlined)

Reply:
A little. Anything containing exception handling won't be inlined. You can use the InliningDiagnoser from benchmarksdotnet to see inlining errors. :)

Reply:
Example using this attribute `[BenchmarkDotNet.Diagnostics.Windows.Configs.InliningDiagnoser(true, true)]`
```
--------------------
WireFormatting_ReadWritePrimitives.WriteDecimal: ShortRun(IterationCount=3, LaunchCount=1, WarmupCount=3)
--------------------
Inliner: Benchmarks.WireFormatting.WireFormatting_ReadWritePrimitives.WriteDecimal - instance int32  ()
Inlinee: System.ThrowHelper.ThrowArgumentOutOfRangeException - void  ()
Fail Reason: does not return
--------------------
Inliner: Benchmarks.WireFormatting.WireFormatting_ReadWritePrimitives.WriteDecimal - instance int32  ()
Inlinee: System.Decimal.GetBits - int32[]  (value class System.Decimal)
Fail Reason: unprofitable inline
--------------------
Inliner: Benchmarks.WireFormatting.WireFormatting_ReadWritePrimitives.WriteDecimal - instance int32  ()
Inlinee: RabbitMQ.Client.Impl.WireFormatting.ThrowWireFormattingException - void  (value class System.Decimal)
Fail Reason: unprofitable inline
--------------------
Inliner: Benchmarks.WireFormatting.WireFormatting_ReadWritePrimitives.WriteDecimal - instance int32  ()
Inlinee: System.ThrowHelper.ThrowArgumentOutOfRangeException - void  ()
Fail Reason: does not return
--------------------
Inliner: Benchmarks.WireFormatting.WireFormatting_ReadWritePrimitives.WriteDecimal - instance int32  ()
Inlinee: System.ThrowHelper.ThrowInvalidTypeWithPointersNotSupported - void  (class System.Type)
Fail Reason: unprofitable inline
--------------------
Inliner: Benchmarks.WireFormatting.WireFormatting_ReadWritePrimitives.WriteDecimal - instance int32  ()
Inlinee: System.ThrowHelper.ThrowArgumentOutOfRangeException - void  (value class System.ExceptionArgument)
Fail Reason: does not return
--------------------
```

Reply:
I'm surprised that it it is faster, not quite get why...
WriteShortstr2 is fully inlined without the extra method.

|         Method | Length |       Mean |     Error |    StdDev | Ratio | RatioSD | Code Size |
|--------------- |------- |-----------:|----------:|----------:|------:|--------:|----------:|
|  WriteShortstr |      0 |  0.7690 ns | 0.0710 ns | 0.0039 ns |  1.00 |    0.00 |     115 B |
| WriteShortstr2 |      0 |  2.6724 ns | 0.1118 ns | 0.0061 ns |  3.48 |    0.02 |     405 B |
|                |        |            |           |           |       |         |           |
|  WriteShortstr |      1 | 12.9989 ns | 0.0236 ns | 0.0013 ns |  1.00 |    0.00 |     364 B |
| WriteShortstr2 |      1 | 13.3813 ns | 0.5709 ns | 0.0313 ns |  1.03 |    0.00 |     405 B |
|                |        |            |           |           |       |         |           |
|  WriteShortstr |     10 | 13.8706 ns | 1.0501 ns | 0.0576 ns |  1.00 |    0.00 |     364 B |
| WriteShortstr2 |     10 | 15.4745 ns | 0.5192 ns | 0.0285 ns |  1.12 |    0.01 |     405 B |
|                |        |            |           |           |       |         |           |
|  WriteShortstr |    100 | 21.8589 ns | 0.5577 ns | 0.0306 ns |  1.00 |    0.00 |     364 B |
| WriteShortstr2 |    100 | 21.0526 ns | 0.1686 ns | 0.0092 ns |  0.96 |    0.00 |     405 B |
|                |        |            |           |           |       |         |           |
|  WriteShortstr |    255 | 28.5284 ns | 0.9557 ns | 0.0524 ns |  1.00 |    0.00 |     364 B |
| WriteShortstr2 |    255 | 29.3146 ns | 4.7681 ns | 0.2614 ns |  1.03 |    0.01 |     405 B |

Reply:
What does WriteShortstr2 look like?

Reply:
Same as the other but without the call to the impl method

Reply:
That doesn't inline actually :) It refuses to do so because of the try/catch which never inlines.

Reply:
Is this still the case in the newer frameworks? (I've not checked the disassembly in that detail.)